go/chisel/server/server_handler.go

```diff
@@ -254,34 +259,72 @@ func (s *Server) handleDynReverse(w http.ResponseWriter, req *http.Request) {
 
 	connectorId := payload.ConnectorID
 	if o, ok := activeTunnels.Load(connectorId); ok {
-		tun := o.(*tunnel.Tunnel)
-		dynPort, err := freeport.GetFreePort()
-		if err != nil {
-			w.WriteHeader(http.StatusInternalServerError)
-			json.NewEncoder(w).Encode(unifiedapiclient.ErrorReply{Status: http.StatusInternalServerError, Message: fmt.Sprintf("Unable to find available port: %s", err)})
-			return
-		}
-		to := payload.To
-		remoteStr := fmt.Sprintf("R:%d:%s", dynPort, to)
-		remote, err := settings.DecodeRemote(remoteStr)
-		if err != nil {
-			w.WriteHeader(http.StatusBadRequest)
-			json.NewEncoder(w).Encode(unifiedapiclient.ErrorReply{Status: http.StatusBadRequest, Message: fmt.Sprintf("The format for the remote (%s) is invalid: %s", to, err)})
-			return
-		}
-
-		remote.Dynamic = true
-		remote.LastTouched = time.Now()
-		settings.ActiveDynReverse.Store(cacheKey, remote)
-		go func() {
-			ctx := context.Background()
```

Comment:
We should simplify this logicl by using context.WithTimeout

ctx, cancel := context.WithTimeout(context.Background(), DYNREVERSE_ERR_WAIT)
defer cancel();

Reply:
It looks like this would close the tunnel after the DYNREVERSE_ERR_WAIT

TBH, I used what I felt the most comfortable with to make sure it's as rock solid as I can given I likely won't be there when this goes onto customer deployments.

I'm happy for you to rework this recipe to something that you feel will be easier to maintain by the team but I'm not sure I'll have the bandwidth to rework this and retest it as thoroughly as I did before submitting the PR.

Reply:
Another approach would have been to set the LocalPort to 0 and let the OS pick the first free port available.
This would simplify the code by having zero need to do a retry.
However, this would require some additional changes as this would require us to return back the port picked from the OS.
I will mark that as future todo.